[fix] Fixed duplicate node ID handling

cestercian opened the pull request:
## **Description**  
This PR addresses [#164](https://github.com/openwisp/netjsongraph.js/issues/164) by improving how `netjsongraph.js` handles duplicate node IDs and expanding test coverage.  

### **Changes**  

- **Duplicate Node Handling:**  
  - `dealJSONData` now filters out duplicate node IDs.  
  - `addData` ensures only unique nodes are rendered.  
  - `mergeData` skips adding duplicate nodes.  

- **Tests & Error Handling:**  
  - Added tests for duplicate ID removal in `dealJSONData`, `mergeData`, and `addData`.  
  - Console warnings notify when duplicate nodes are detected.  

- **Refactoring:**  
  - Improved object handling with utility functions.  
  - Organized test data using `Map` for better maintainability.  

These changes improve stability and reliability.  

## **Checklist**  
- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).  
- [x] I have manually tested these changes.  
- [x] I have written/updated test cases.  
- [x] I have updated the documentation (if applicable).  

## **Reference to Existing Issue**  
Closes [#164](https://github.com/openwisp/netjsongraph.js/issues/164).  

## **Screenshots**  
<img width="1710" alt="image" src="https://github.com/user-attachments/assets/da6f9da4-483d-4950-b08c-87974b4e51ba" />

cestercian commented:
Fixed code style issues in `src/js/netjsongraph.render.js`, 
`src/js/netjsonWorker.js`, and `test/netjsongraph.duplicateNodes.test.js`
using Prettier. Ensures consistency in formatting.

cestercian commented:
**Fix QA Checks & Lint Errors**  

- Fixed Prettier formatting issues.  
- Resolved ESLint errors by removing unused variables.  
- Updated commit message to follow `[feature/fix/change]` format.  
- Reran `./run-qa-checks`, and the build has passed.  

Ready for review. Let me know if any changes are needed.

nemesifier reviewed (changes requested):
The logic which is dealing with duplicate nodes is being added to 3 different files in 3 different ways. 

Too much redundancy, it will be a burden to maintain.
Please make sure the logic is implemented only one time.
Do we really need to check for this node ID duplication in 3 different places? If this is really necessary, we can resolve it with a reusable function which is called in different places. 
As of now, it seems too much code now to handle a corner case.

cestercian commented:
<img width="1687" alt="image" src="https://github.com/user-attachments/assets/325fd3fe-bfd9-4824-8bb4-568ef29adde1" />


### Refactor node deduplication logic

- Added `deduplicateNodesById` utility in `netjsonWorker.js` for consistent deduplication.  
- Refactored `dealJSONData`, `addData`, and `mergeData` to use the utility.  
- Updated tests to mock `deduplicateNodesById`.

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+
+    // Verify that the duplicate node was not added
+    expect(self.data.nodes.filter((node) => node.id === "node1").length).toBe(
+      expected.node1Count,
```
This fails. However, the comment says "Verify that the duplicate node was not added", aren't we testing that the duplicate node is added once and any duplicate is ignored? Maybe the comment is slightly misleading?

cestercian replied:
Yeah, you’re absolutely right — that comment does sound misleading.

test is supposed to ensure that a duplicate node isn’t added multiple times. I'll update the comment

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+
+describe("NetJSONGraph Duplicate Node ID Handling", () => {
+  // Test the enhanced dealJSONData function
+  test("dealJSONData should handle duplicate node IDs gracefully", () => {
```
This test is not failing at all in master, have you tried it? I would be expecting the test to fail since AFAIK the library crashes if there's duplicate nodes. This test doesn't seem to be effective in replicating the bug. Please double check on your end.

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+};
+
+describe("NetJSONGraph Duplicate Node ID Handling", () => {
+  // Test the enhanced dealJSONData function
```
These comments before the test definition are redundant. Please remove.

cestercian replied:
sure !

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+    expect(processedData.flatNodes.node2).toBeDefined();
+  });
+
+  // Test the enhanced mergeData function
```
These comments before the test definition are redundant. Please remove.

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+    expect(self.data.links.length).toBe(expected.finalLinkCount);
+  });
+
+  // Test the addData function's safeguard
```
These comments before the test definition are redundant. Please remove.

nemesifier commented on test/netjsongraph.duplicateNodes.test.js:
```diff
@@ -0,0 +1,247 @@
+    netJSONGraphRender.addData(newData, self);
+
+    // Verify that the duplicate node was removed by the safeguard
+    expect(self.data.nodes.filter((node) => node.id === "node1").length).toBe(
```
This is also failing so it should be good.

nemesifier reviewed (changes requested):
I copied the test file on master to verify how the tests fail and noticed only two failures. See my comments below.

cestercian commented:
No, it's not failing on master, because the purpose of the `dealJSONData` function is to sanitize the initial input by **deduplicating nodes based on their `id`**.

The test case **"dealJSONData should handle duplicate node IDs gracefully"** is meant to validate that behavior, so it's expected to pass even when the input contains duplicates — since those duplicates should be removed during processing.

cestercian commented:
The version on master doesn't crash when handling duplicate node IDs — but it also doesn't handle them correctly. What happens is:

if we pass duplicates needs in an array.
In flatNodes, which uses node.id as a key, the last duplicate silently overwrites the earlier one.
So while it doesn’t throw an error, it silently corrupts the node data, which leads to issues like incorrect rendering or unexpected behavior.

nemesifier reviewed (approved):
Thanks @cestercian :+1:
